kroxylicious-filters/kroxylicious-oauthbearer-validation/src/main/java/io/kroxylicious/proxy/filter/oauthbearer/OauthBearerValidationFilter.java

```diff
@@ -0,0 +1,207 @@
+                    .exceptionallyCompose(e -> {
+                        if (e.getCause() instanceof SaslAuthenticationException) {
+                            SaslAuthenticateResponseData failedResponse = new SaslAuthenticateResponseData()
+                                    .setErrorCode(SASL_AUTHENTICATION_FAILED.code())
+                                    .setErrorMessage(e.getMessage())
+                                    .setAuthBytes(request.authBytes());
+                            LOGGER.debug("SASL Authentication failed : {}", e.getMessage(), e);
+                            return context.requestFilterResultBuilder().shortCircuitResponse(failedResponse).withCloseConnection().completed();
+                        }
+                        else {
+                            LOGGER.debug("SASL error : {}", e.getMessage(), e);
+                            return context.requestFilterResultBuilder()
+                                    .shortCircuitResponse(
+                                            new SaslAuthenticateResponseData()
+                                                    .setErrorCode(UNKNOWN_SERVER_ERROR.code())
+                                                    .setAuthBytes(request.authBytes()))
+                                    .withCloseConnection()
+                                    .completed();
+                        }
+                    });
+        }
+    }
+
+    @Override
+    public CompletionStage<ResponseFilterResult> onSaslAuthenticateResponse(short apiVersion, ResponseHeaderData header,
+                                                                            SaslAuthenticateResponseData response, FilterContext context) {
+        if (response.errorCode() == NONE.code()) {
+            this.validateAuthentication = false;
+        }
+        return context.forwardResponse(header, response);
+    }
+
+    private CompletionStage<byte[]> authenticate(byte[] authBytes) {
+        String digest;
+        try {
+            digest = digestBytes(authBytes);
+        }
+        catch (NoSuchAlgorithmException e) {
+            return CompletableFuture.failedStage(e);
+        }
+        Duration delay = strategy.getDelay(rateLimiter.get(digest).get());
+        return schedule(() -> {
+            try {
+                return CompletableFuture.completedStage(doAuthenticate(authBytes));
+            }
+            catch (Exception e) {
+                return CompletableFuture.failedStage(e);
+            }
+        }, delay)
+                .whenComplete((bytes, e) -> {
+                    if (e != null) {
+                        rateLimiter.get(digest).incrementAndGet();
+                    }
+                    else {
+                        rateLimiter.invalidate(digest);
+                    }
+                });
+    }
+
+    private byte[] doAuthenticate(byte[] authBytes) throws SaslException {
+        try {
+            return this.saslServer.evaluateResponse(authBytes);
```

Comment:
Is the error handling right? `org.apache.kafka.common.security.oauthbearer.internals.OAuthBearerSaslServer#process` will return a `jsonErrorResponse` if the token is null.    That's not an exception.    I think you should be treating cases where complete = false (or evaluateResponse return something other than an empty byte array as an error.